csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -309,29 +298,71 @@ private OutputStream StartBinaryContext()
         private protected void FinishEncapsulation(OutputStream.Position encapsulationEnd) =>
             _encapsulationEnd = encapsulationEnd;
 
-        // Finish prepare the frame for sending, adjust the last written segment to match the offset of the written
-        // data, if the frame contains a binary context we rewrite the size.
+        // Finish prepares the frame for sending, adjusts the last written segment to match the offset of the written
+        // data. If the frame contains a binary context, Finish appends the entries from _defaultBinaryContext (if any)
+        // and rewrites the binary context dictionary size.
         internal virtual void Finish()
         {
             if (!IsSealed)
             {
                 if (_binaryContextOstr is OutputStream ostr)
                 {
-                    Debug.Assert(_encapsulationEnd != null);
-                    ostr.RewriteFixedLengthSize20(_binaryContextKeys.Count,
-                                                  _encapsulationEnd.Value,
-                                                  2);
+                    Debug.Assert(_binaryContextKeys != null);
                     Data[^1] = Data[^1].Slice(0, ostr.Tail.Offset);
+                    if (_defaultBinaryContext.Count > 0)
+                    {
+                        // Add segment for each slot that was not written yet.
+                        var istr = new InputStream(_defaultBinaryContext, Encoding.V2_0);
+                        int dictionarySize = istr.ReadSize();
+                        for (int i = 0; i < dictionarySize; ++i)
+                        {
+                            int key = istr.ReadVarInt();
+                            int entrySize = istr.ReadSize();
+                            if (ContainsKey(key))
+                            {
+                                istr.Skip(entrySize);
+                            }
+                            else
+                            {
+                                Data.Add(_defaultBinaryContext.Slice(istr.Pos, entrySize));
+                                AddKey(key);
+                            }
+                        }
+                    }
+
+                    Debug.Assert(_encapsulationEnd != null);
+                    ostr.RewriteFixedLengthSize20(_binaryContextKeys.Count, _encapsulationEnd.Value, 2);
                 }
                 else
                 {
-                    Debug.Assert(_encapsulationEnd != null);
-                    Data[^1] = Data[^1].Slice(0, _encapsulationEnd.Value.Offset);
+                    Debug.Assert(_binaryContextKeys == null);
+                    if (_defaultBinaryContext.Count > 0) // only when forwarding an ice2 request or response
```

Comment:
can we avoid this nested if

```
else if (_defaultBinaryContext.Count > 0)
{
....
}
else
{
...
}
```

And just get rid of `Debug.Assert(_binaryContextKeys == null);`

Reply:
I find the current nested version clearer.